lib/package_config.dart

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// A package configuration is a way to assign file paths to package URIs,
+/// and vice-versa,
+library package_config.package_config;
+
+import "dart:io" show File, Directory;
+import "src/discovery.dart" as discover;
+import "src/package_config.dart";
+import "src/package_config_json.dart";
+
+export "src/package_config.dart" show PackageConfig, Package;
+export "src/discovery_analysis.dart" show PackageContext;
+
+/// Reads a specific package configuration file.
+///
+/// The file must exist and be readable.
+/// It must be either a valid `package_config.json` file
+/// or a valid `.packages` file.
+/// It is considered a `package_config.json` file if its first character
+/// is a `{`.
+///
+/// If the file is a `.packages` file, first checks if there is a
+/// `.dart_tool/package_config.json` file next to the original file,
+/// and if so, loads that instead.
+///
+/// If [extraData] is provided, and the loaded file is a `package_config.json`
+/// file, then any unknown JSON object entries in the main JSON object of the
+/// file are added to [extraData] with the same name.
+PackageConfig loadPackageConfig(File file,
```

Comment:
I ... do not like the "Sync" suffix. At all. (And it should not be abbreviated).

I prefer to be opinionated and have just one version, and here I think synchronous is fine. It's not a large file.

One option is to provide a *parse* function instead of a *load* function, so users can load their bytes any way they want to. The issue with that is that the function actually tries to load two files if the first file turns out to be a `.packages` file. I want to control that logic so the user doesn't forget to do that.

Reply:
OK, I've made the API async. There's no easy way to make it both, so it'll all be async, and code using it will have to be so too.

Reply:
fwiw I would also be in favor of a `parse` api because that would make using this package in tests easier

Reply:
A "parse" thing is needed. In the CFE we basically cannot provide you a "File".

Reply:
The API now allows the load functions to provide an optional `loader` function which abstracts over fetching the data. That will allow you to programmatically provide the data without going to the file system.